data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsDispatcher.java

```diff
@@ -0,0 +1,163 @@
+
+        /**
+         * Current implementation, timestamp is generated by system time during transmission.
+         * To properly extract timestamp we need to order the InputLogEvents. Can be done by
+         * refactoring buffer class with timestamp param, or adding a sorting algorithm in between
+         * making the PLE object (in prepareInputLogEvents).
+         */
+
+        for (byte[] data : eventMessageBytes) {
+            InputLogEvent tempLogEvent = InputLogEvent.builder()
+                    .message(new String(data, StandardCharsets.UTF_8))
+                    .timestamp(System.currentTimeMillis())
+                    .build();
+            logEventList.add(tempLogEvent);
+        }
+
+        return logEventList;
+    }
+
+    public void dispatchLogs(List<InputLogEvent> inputLogEvents, Collection<EventHandle> eventHandles) {
+        PutLogEventsRequest putLogEventsRequest = PutLogEventsRequest.builder()
+                .logEvents(inputLogEvents)
+                .logGroupName(logGroup)
+                .logStreamName(logStream)
+                .build();
+
+        executor.execute(Uploader.builder()
+                .cloudWatchLogsClient(cloudWatchLogsClient)
+                .cloudWatchLogsMetrics(cloudWatchLogsMetrics)
+                .putLogEventsRequest(putLogEventsRequest)
+                .eventHandles(eventHandles)
+                .backOffTimeBase(backOffTimeBase)
+                .retryCount(retryCount)
+                .build());
+    }
+
+    @Builder
+    protected static class Uploader implements Runnable {
+        private final CloudWatchLogsClient cloudWatchLogsClient;
+        private final CloudWatchLogsMetrics cloudWatchLogsMetrics;
+        private final PutLogEventsRequest putLogEventsRequest;
+        private final Collection<EventHandle> eventHandles;
+        private final int retryCount;
+        private final long backOffTimeBase;
+
+        @Override
+        public void run() {
+            upload();
+        }
+
+        public void upload() {
+            boolean failedToTransmit = true;
+            int failCount = 0;
+
+            try {
+                while (failedToTransmit && (failCount < retryCount)) {
+                    try {
+                        cloudWatchLogsClient.putLogEvents(putLogEventsRequest);
+
+                        cloudWatchLogsMetrics.increaseRequestSuccessCounter(1);
+                        failedToTransmit = false;
+
+                    } catch (CloudWatchLogsException | SdkClientException e) {
+                        LOG.error("Failed to push logs with error: {}", e.getMessage());
+                        cloudWatchLogsMetrics.increaseRequestFailCounter(1);
+                        Thread.sleep(calculateBackOffTime(backOffTimeBase, failCount));
+                        failCount++;
+                    }
+                }
+            } catch (InterruptedException e) {
+                LOG.warn("Got interrupted while waiting!");
```

Comment:
Better log needed here. This doesn't tell the customer anything. Interrupted by what? stacktrace? while waiting for what? in what class?

Reply:
Thank you, I have added a more appropriate message alongside the actual thrown error message.